document/js-api/index.bs

```diff
@@ -1138,47 +1140,35 @@ The algorithm <dfn>ToWebAssemblyValue</dfn>(|v|, |type|) coerces a JavaScript va
-            1. If |testresult| is true,
-                1. Return [=ref.func=] |funcaddr|.
-        1. Throw a {{TypeError}}.
-    1. If |heaptype| matches [=any=],
-        1. If |v| [=is a Number=],
-            1. Let |i32| be ? [$ToInt32$](|v|).
-            1. If |v| is equal to |i32| and [=ℝ=](|v|) < 2<sup>30</sup> and [=ℝ=](|v|) ⩾ -2<sup>30</sup>,
-                1. If [=i31=] does not match |heaptype|,
-                    1. Throw a {{TypeError}}.
-                1. Return [=ref.i31=] |v|.
-        1. If |v| is an [=Exported GC Object=],
-            1. Let |objectaddr| be the value of |v|'s \[[ObjectAddress]] internal slot.
-            1. Let |objectkind| be the value of |v|'s \[[ObjectKind]] internal slot.
-            1. If |objecttype| is "array",
-                1. If [=array=] does not match |heaptype|,
-                    1. Throw a {{TypeError}}.
-                1. Return [=ref.array=] |objectaddr|.
-            1. If |objectkind| is "struct",
-                1. If [=struct=] does not match |heaptype|,
-                    1. Throw a {{TypeError}}.
-                1. Return [=ref.struct=] |objectaddr|.
-        1. If [=any=] does not match |heaptype|,
-            1. Throw a {{TypeError}}.
+        1. Let |r| be [=ref.null=] |heaptype|.
+    1. Else if |v| [=is a Number=] and |v| is equal to ? [$ToInt32$](|v|) and [=ℝ=](|v|) < 2<sup>30</sup> and [=ℝ=](|v|) ⩾ -2<sup>30</sup>,
+        1. Let |r| be [=ref.i31=] |v|.
+    1. Else if |v| is an [=Exported Function=],
+        1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
+        1. Let |r| be [=ref.func=] |funcaddr|.
+    1. Else if |v| is an [=Exported GC Object=],
+        1. Let |objectaddr| be the value of |v|'s \[[ObjectAddress]] internal slot.
+        1. Let |objectkind| be the value of |v|'s \[[ObjectKind]] internal slot.
+        1. If |objectkind| is "array",
+            1. Let |r| be [=ref.array=] |objectaddr|.
+        1. If |objectkind| is "struct",
+            1. Let |r| be [=ref.struct=] |objectaddr|.
+    1. Else,
         1. Let |map| be the [=surrounding agent=]'s associated [=host value cache=].
         1. If a [=host address=] |hostaddr| exists such that |map|[|hostaddr|] is the same as |v|,
             1. Return [=ref.host=] |hostaddr|.
         1. Let [=host address=] |hostaddr| be the smallest address such that |map|[|hostaddr|] [=map/exists=] is false.
         1. [=map/Set=] |map|[|hostaddr|] to |v|.
-        1. Return [=ref.host=] |hostaddr|.
-    1. If |heaptype| matches [=extern=],
-        1. Let |r| be [=ToWebAssemblyValue=](|v|, [=ref=] |null| [=any=]).
-        1. Return [=ref.extern=] |r|.
+        1. Let |r| be [=ref.host=] |hostaddr|.
+    1. Let |store| be the current agent's [=associated store=].
+    1. Let |actualtype| be [=ref_type=](|store|, |r|).
+    1. If [=match_valtype=](|type|, [=ref=] |null| [=heap-type/func=]) is false,
+        1. Throw a {{TypeError}}.
```

Comment:
Is this supposed to use `actualtype` somewhere? Also, why is the store involved?

Reply:
Oops, yes, forgot to update that line after copy & paste. Fixed.

The store is needed to obtain the precise type of a reference.